### PR TITLE
Extracted functions to static methods

### PR DESCRIPTION
**Pull request by GrahamCampbell**
This PR pairs with https://github.com/guzzle/guzzle/pull/2712. After this is merged, we can tag 1.7.0.

### TODO:

- [x] Merge CI and HHVM fixes https://github.com/guzzle/psr7/pull/347
- [x] Extract all functions to static class methods
- [x] Deprecate all functions and point to their replacements
- [x] Replace all uses in src
- [x] Update the tests
- [x] Update documentation and examples
- [x] Update change log

### Mappings:

| Original Function | Replacement Method |
|----------------|----------------|
| `str` | `Message::toString` |
| `uri_for` | `Utils::uriFor` |
| `stream_for` | `Utils::streamFor` |
| `parse_header` | `Header::parse` |
| `normalize_header` | `Header::normalize` |
| `modify_request` | `Utils::modifyRequest` |
| `rewind_body` | `Message::rewindBody` |
| `try_fopen` | `Utils::tryFopen` |
| `copy_to_string` | `Utils::copyToString` |
| `copy_to_stream` | `Utils::copyToStream` |
| `hash` | `Utils::hash` |
| `readline` | `Utils::readLine` |
| `parse_request` | `Message::parseRequest` |
| `parse_response` | `Message::parseResponse` |
| `parse_query` | `Query::parse` |
| `build_query` | `Query::build` |
| `mimetype_from_filename` | `MimeType::fromFilename` |
| `mimetype_from_extension` | `MimeType::fromExtension` |
| `_parse_message` | `Message::parseMessage` |
| `_parse_request_uri` | `Message::parseRequestUri` |
| `get_message_body_summary` | `Message::bodySummary` |
| `_caseless_remove` | `Utils::caselessRemove` |

NB The overall idea of these changes was approved by the original author, Michael: https://github.com/guzzle/promises/pull/108.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Friendly ping @Tobion and @Nyholm. It'd be good to have this merged so I can focus on PHP 8.0 support next, without conflicts.

File: .travis.yml [hunk 1]
```diff
@@ -1,31 +1,39 @@
 language: php
 
-php:
```
Comment (Nyholm) [inline .travis.yml:3]: We could rebase this PR now when #347 is merged

File: src/Utils.php [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    }
+
+    /**
+     * Create a new stream based on the input type.
+     *
+     * Options is an associative array that can contain the following keys:
+     * - metadata: Array of custom metadata.
+     * - size: Size of the stream.
+     *
+     * This method accepts the following `$resource` types:
```
Comment (Nyholm) [inline src/Utils.php:267]: Nice!

File: src/functions.php [hunk 3]
```diff
@@ -75,86 +48,57 @@ function uri_for($uri)
  * - metadata: Array of custom metadata.
  * - size: Size of the stream.
  *
+ * This method accepts the following `$resource` types:
```
Comment (Nyholm) [inline src/functions.php:51]: Not sure we need to add this on a deprecated function

File: src/Message.php [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+     *
+     * The array contains the "start-line" key containing the start line of
+     * the message, "headers" key containing an associative array of header
+     * array values, and a "body" key containing the body of the message.
+     *
+     * @param string $message HTTP request or response to parse.
+     *
+     * @return array
+     */
+    public static function parseMessage($message)
```
Comment (Nyholm) [inline src/Message.php:118]: This should be internal as `_parse_message`.

File: src/Utils.php [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+final class Utils
+{
+    /**
+     * Remove the items given by the keys, case insensitively from the data.
+     *
+     * @param iterable<string> $keys
+     *
+     * @return array
+     */
+    public static function caselessRemove($keys, array $data)
```
Comment (Nyholm) [inline src/Utils.php:19]: This should be marked as internal too.

File: src/Message.php [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+
+    /**
+     * Constructs a URI for an HTTP request message.
+     *
+     * @param string $path    Path from the start-line
+     * @param array  $headers Array of headers (each value an array).
+     *
+     * @return string
+     */
+    public static function parseRequestUri($path, array $headers)
```
Comment (Nyholm) [inline src/Message.php:181]: Same here, mark it as internal

File: README.md [hunk 7]
```diff
@@ -297,235 +297,290 @@ stream from a PSR-7 stream.
...
 
-There are various functions available under the `GuzzleHttp\Psr7` namespace.
+There are various static methods available under the `GuzzleHttp\Psr7` namespace.
 
 
-## `function str`
+## `GuzzleHttp\Psr7\Message::toString`
 
-`function str(MessageInterface $message)`
+`toString(MessageInterface $message): string`
```
Comment (Nyholm) [inline README.md:313]: To be consistent: 
 
```suggestion
`public static function toString(MessageInterface $message): string`
```

**Review by Nyholm — CHANGES_REQUESTED** [review]
Thank you. I like this. 

I just had some minors. After that Im 👍

File: src/functions.php (hunk 3 above)
Comment (GrahamCampbell) [reply src/functions.php:51]: I took the view that the functions and static methods should have identical phpdoc.

**Review by Nyholm — APPROVED** [review]
Thank you

**Comment by GrahamCampbell** [comment]
🚀

**Pull request merged**